modules/targetAcquisition/taxi/Taxi.py

```diff
@@ -0,0 +1,331 @@
+        self.distanceFromBox = distanceFromBox
+        self.minDistanceFromBox = minDistanceFromBox
+        self.moveWaitTarget = moveWaitTarget
+        self.recalibrate = recalibrate
+        self.lastBbox = lastBbox
+
+        # Instead of keeping the following as local variables in the while loop,
+        # they become object attributes and get updated each time Taxi.main() is called
+        self.frameCount = frameCount
+        self.totalWait = totalWait
+        self.moveWaitCount = moveWaitCount
+        self.distanceCount = distanceCount
+
+
+    def find_overlapped_bbox(self):
+        """
+        Given the last known bbox and a list of new ones, return the new bbox which overlaps the old box the most.
+        Helps recalibrate the tracked bbox to fit the cardboard box better, so distance calculation is more accurate.
+        """
+        # If the new scan found zero bounding boxes, return no bbox
+        if len(self.bbox) < 1 or len(self.lastBbox) < 1:
+            print("ERROR: box list too short. Aborting")
+            return None, 0
+
+        # Area of the tracked bbox
+        last = self.lastBbox[0]
+        originalArea = (last[1][1] - last[0][1]) * (last[1][0] - last[0][0])
+
+        bbox = self.bbox[0]
+        maxOverlap = 0
+        for box in self.bbox:
+            x0 = max(last[0][0], box[0][0])
+            x1 = min(last[1][0], box[1][0])
+            y0 = max(last[0][1], box[0][1])
+            y1 = min(last[1][1], box[1][1])
+            newArea = (y1 - y0) * (x1 - x0)
+            overlap = newArea / originalArea
+
+            if overlap > maxOverlap:
+                maxOverlap = overlap
+                bbox = box
+
+        # If even the best match is not good enough, return no bboxes
+        if maxOverlap < 0.75:
+            return None, maxOverlap
+
+        # Return the bbox with max overlap
+        print(f"Matched bbox: {bbox}, max overlap: {maxOverlap}")
+        return bbox, maxOverlap
+
+    def calculate_distance(self, pts):
+        """
+        Calculate approximate distance between box and drone
+        """
+
+        # Calculating object height in pixels by extracting y coordinates from each tuple 'pts'
+        objectHeight = pts[0][1] - \
+            pts[1][1] if (pts[0][1] > pts[1][1]) else pts[1][1] - pts[0][1]
+
+        # Calculate distance
+        distance = (FOCAL_LENGTH * REAL_HEIGHT * IMAGE_HEIGHT) / \
+            (objectHeight * SENSOR_HEIGHT)
+
+        return distance
+
+    def set_state(self, state):
+        """
+        Prepare variables for box detection, tracking etc.
+
+        Parameters
+        ----------
+        state: string
+            state of object recognition ("BOX", "TRACK", ...)
+        """
+        if state == "BOX" or state == 0:
+            self.state = "BOX"
+
+        elif state == "TRACK" or state == 1:
+            self.state = "TRACK"
+
+            # If switching to track from box for the first time, use the box ID
```

Comment:
Random indent

Reply:
Assuming there was supposed to be an else here?

Reply:
No, this is correct, its just not very readable due to the space. this is all tracking logic.